article/article.md

```diff
@@ -0,0 +1,153 @@
+
+Around this time, I began thinking about other ways to render our map tiles. One option was to create our own tile server to do the colorization. A custom server could be better suited to our specific needs, and, crucially, we'd be more able optimize its performance because we'd understand it better. Yet a server-side solution had some inherent limitations. Rendering a large number of pixels on the CPU is bound to be slow, because the pixels must be processed sequentially (or with limited parallelization taking advantage of multicore architecture). With commodity hardware and the VMs in which our containerized services run, there would be hard limits on how fast the tiles could be produced. It seemed to me that the ideal way to render the images was on the [GPU](https://en.wikipedia.org/wiki/Graphics_processing_unit), which is specifically designed to process large numbers of pixels in parallel. Fortunately, modern browsers provide a way to do GPU processing: [WebGL](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API)! A WebGL solution would mean the tile server would have very little to do - just retrieve pixel data from PostGIS and then pass it in essentially raw form to the client. The user's browser would then be responsible for actually colorizing the pixels, and it could take advantage of GPU rendering to do this very efficiently.
+
+Having the raw floating-point pixel values available clientside would have another advantage as well - allowing the user quick access to the actual pixel-level estimates in the visualization. While humans generally have a pretty keen color sense, the result of mentally translating a color from a gradient back to the value it's supposed to represent seemed pretty imprecise. Our visualization already had the capability of getting the data value of a given pixel when clicked by the user, but doing so required a backend request and was therefore relatively slow. Having this data available clientside would mean we could retrieve and display these values very quickly.
+
+## Problem 1: Reading floats from a texture in a WebGL shader
+
+The first and most important hurdle I foresaw with this new approach was a limitation of WebGL 1.0: the lack of support for floating-point textures. A "texture," in graphics parlance, is essentially an image you pass to the GPU. Once a texture is loaded, you can read any pixel or pixels you like from it in your shader, which is a small user-defined program that runs on the GPU. WebGL 1.0, by far the most widely implemented version of the standard, allows you to create textures using a variety of pixel types, like RGB and RGBA, but unless the extension `OES_texture_float` is enabled for a user's hardware and browser, it's not possible to use a texture in which the pixels represent floats. WebGL 2.0 adds support for float textures, but our analytics indicated that many of the users of our tool didn't have the capability of using WebGL 2.0.
```

Comment:
"...you can read any pixel or pixels you like from it in your shader", don't use "it" here: "you can read any pixel or pixels you like from the texture in your shader...".

Reply:
revised the wording here; let me know if it's clearer